File: ios/main/StallionDownloadCacheManager.swift
import Foundation

@objc class StallionDownloadCacheManager: NSObject {
    
    private static let metaFileName = "download-cache.meta"
    
    @objc static func getDownloadCache(config: StallionConfig, downloadUrl: String, downloadPath: String) -> Int64 {
        let lastDownloadingUrl = config.lastDownloadingUrl ?? ""
        let alreadyDownloaded = readMetaFile(at: downloadPath)
        
        if lastDownloadingUrl != downloadUrl || alreadyDownloaded <= 0 {
          config.updatePrevDownloadUrl(downloadUrl)
          StallionFileManager.deleteFileOrFolderSilently(downloadPath)
            return 0
        } else {
            return alreadyDownloaded
        }
    }
    
    @objc static func saveDownloadCache(path: String, bytes: Int64) {
        let metaPath = (path as NSString).appendingPathComponent(metaFileName)
        do {
            try "\(bytes)".data(using: .utf8)?.write(to: URL(fileURLWithPath: metaPath))
        } catch {
            // Silently ignore
        }
    }
    
    private static func readMetaFile(at path: String) -> Int64 {
        let metaPath = (path as NSString).appendingPathComponent(metaFileName)
        let fileURL = URL(fileURLWithPath: metaPath)
        
        guard FileManager.default.fileExists(atPath: metaPath) else { return 0 }
        
        do {
            let data = try Data(contentsOf: fileURL)
            let stringValue = String(data: data, encoding: .utf8) ?? "0"
            return Int64(stringValue.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
        } catch {
            return 0
        }
    }
  
    @objc static func deleteDownloadCache(path: String) {
        let metaPath = (path as NSString).appendingPathComponent("download-cache.meta")
        let fileURL = URL(fileURLWithPath: metaPath)
        try? FileManager.default.removeItem(at: fileURL)
    }
  
}

File: android/src/main/java/com/stallion/networkmanager/StallionDownloadCacheManager.java
package com.stallion.networkmanager;

import com.stallion.storage.StallionConfig;
import com.stallion.utils.StallionFileManager;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.Objects;

public class StallionDownloadCacheManager {
  private static final String metaFilePath = "/download-cache.meta";

  public static long getDownloadCache(StallionConfig config, String downloadUrl, String downloadPath) {
    String lastDownloadingUrl = config.getLastDownloadingUrl();
    long alreadyDownloaded = readMetaFile(downloadPath);
    if(!Objects.equals(lastDownloadingUrl, downloadUrl) || alreadyDownloaded <= 0) {
      config.setLastDownloadingUrl(downloadUrl);
      StallionFileManager.deleteFileOrFolderSilently(new File(downloadPath));
      return 0;
    } else {
      return alreadyDownloaded;
    }
  }

  public static void saveDownloadCache(String path, long bytes) {
    try (FileOutputStream fos = new FileOutputStream(path + metaFilePath)) {
      fos.write(Long.toString(bytes).getBytes());
    } catch (Exception ignored) {}
  }

  private static long readMetaFile(String path) {
    File meta = new File(path + metaFilePath);
    if (!meta.exists()) return 0;
    try (FileInputStream fis = new FileInputStream(meta)) {
      byte[] data = new byte[(int) meta.length()];
      fis.read(data);
      return Long.parseLong(new String(data));
    } catch (Exception e) {
      return 0;
    }
  }

  public static void deleteDownloadCache(String path) {
    File meta = new File(path + metaFilePath);
    if (meta.exists()) meta.delete();
  }
}
